lalrpop/src/normalize/macro_expand/mod.rs

```diff
@@ -52,21 +52,35 @@ impl MacroExpander {
         }
     }
 
-    fn expand(&mut self, items: &mut Vec<GrammarItem>) -> NormResult<()> {
-        let mut counter = 0;
+    fn expand(&mut self, items: &mut Vec<GrammarItem>, recursion_limit: u16) -> NormResult<()> {
+        let mut counter = 0; // Number of items
+        let mut loop_counter = 0;
+
         loop {
             // Find any macro uses in items added since last round and
             // replace them in place with the expanded version:
             for item in &mut items[counter..] {
                 self.replace_item(item);
             }
             counter = items.len();
+            loop_counter += 1;
 
             // No more expansion to do.
             if self.expansion_stack.is_empty() {
                 return Ok(());
             }
 
+            if loop_counter > recursion_limit {
+                // Too much recursion
+                // We know unwrap() is safe, because we just checked is_empty()
+                let sym = self.expansion_stack.pop().unwrap();
+                return_err!(
+                            sym.span,
+                            "Exceeded recursion cap ({}) while expanding this macro.  This typically is a symptom of infinite recursion during macro resolution.  If you believe the recursion will complete eventually, you can increase this limit using Configuration::set_macro_recursion_limit().",
+                            recursion_limit
+                        );
+            }
```

Comment:
I don't know if it's worth it, but note that you can get rid of the unwrap by grouping both checks in one pattern matching, something like:

```rust
match self.expansion_stack.pop() {
  // No more expansion to do.
  None => return Ok(()),
  // Too much recursion
  Some(sym) if loop_counter > recursion_limit => { ...code path of the if...},
  _ => (),
}
```

Reply:
Yeah, I did consider that, but I do think that the code is more readable keeping these as separate.  In my mind at least, these are logically two separate checks.

1. Are we all done?
2. Have we exceeded the recursion limit?

The fact that check 2 implicitly needs info about check 1 is really just a detail, not a fundamental bit of code logic.

Also, I think the code as you've written it has a bug - if `pop()` returns a value, and we haven't exceeded the limit, then this would discard that value, when instead we want to use it in the ensuing `while` loop.  It's a fixable bug, but fixing it would further complicate the code.

Reply:
I agree that the current code is more readable, however, sometimes people like to minimize the number of `unwrap` by helping the compiler with proving that indeed this doesn't need to be there. So it's really a trade-off. I personally have no strong opinion, so I'll let you pick the solution you think is better :slightly_smiling_face:

Reply:
My opinion remains unchanged from above :)  Since you left it to me, I'm going to go ahead and merge this as is.  Thanks for the review!